### PR TITLE
[RDY] Remove jade checker and replace it with pug checker

### LINKED ISSUE #951: jade is now called pug  (trademark problem)-- we should probably fix it so flycheck works
I'm gonna take this on but is there a way to handle this in a way that doesn't break the current jade support?

### PR DESCRIPTION
**Pull request by robbyoconnor**
This removes the jade checker and adds the a checker for pug. 

Closes GH-951

### CODE REVIEW

**Comment by robbyoconnor** [comment]
@flycheck/maintainers @lunaryorn

**Comment by cpitclaudel** [comment]
Thanks, looks great :) (LGTM) Could you add a news entry, too?

**Comment by cpitclaudel** [comment]
(maybe the news entry should mention that the patterns have been updated, too?)

**Comment by robbyoconnor** [comment]
I'll fix the formatting snafu hold on

**Comment by robbyoconnor** [comment]
I'll fix that. Gotta let them know that this won't work as a drop-in for Jade.

**Comment by robbyoconnor** [comment]
I added `f` as a dependency in the `Cask` file. It broke `make format`

**Comment by robbyoconnor** [comment]
Hold off -- I'm gonna keep the jade checker -- hold on.

File: Cask [hunk 1]
```diff
@@ -19,13 +19,13 @@
  (depends-on "d-mode")
  (depends-on "erlang")
  (depends-on "ess")
+ (depends-on "f")
```
Comment (lunaryorn) [inline Cask]: Please remove this line.  Look above, `f` is already there…

File: flycheck.el [hunk 2]
```diff
@@ -7784,6 +7771,21 @@ See https://github.com/processing/processing/wiki/Command-Line"
...
+
+See URL `https://pugjs.org/'."
+  :command ("pug")
+  :standard-input t
+  :error-patterns
+  ((error line-start
+          "Error: Pug:" line ":" column (zero-or-more not-newline) "\n"
+          (one-or-more (or (zero-or-more not-newline) "|" (zero-or-more not-newline) "\n")
+                       (zero-or-more "-")  (zero-or-more not-newline) "|" (zero-or-more not-newline) "\n")
+          (zero-or-more not-newline) "\n" (one-or-more  (zero-or-more not-newline) "|" (zero-or-more not-newline) "\n") (zero-or-more not-newline) "\n"
```
Comment (lunaryorn) [inline flycheck.el]: Please wrap lines longer than 80 characters.

File: test/flycheck-test.el [hunk 3]
```diff
@@ -3282,10 +3282,6 @@ Why not:
    '(8 5 warning "discarding unexpected <spam>"
        :checker html-tidy)))
 
-(flycheck-ert-def-checker-test jade jade nil
-  (flycheck-ert-should-syntax-check
-   "language/jade.jade" 'jade-mode
```
Comment (lunaryorn) [inline test/flycheck-test.el]: Did you remove the jade file as well?

File: CHANGES.rst [hunk 4]
```diff
@@ -157,6 +158,7 @@ version number.  Breaking changes may occur at any point.
     with ``go vet`` [GH-765] [GH-897]
   - Add ``flycheck-ghc-stack-use-nix`` to enable Nix support for Stack GHC
     [GH-913]
+  - Removal of the Jade checker [GH-951] [GH-1084]
```
Comment (lunaryorn) [inline CHANGES.rst]: That's the wrong version to add the note to.  Please rebase onto `master` and add the change notes to version 30 which is currently in development.

Furthermore, please keep the grammar and phrasing consistent to the other entries, i.e. not "Removal of", but just "remove".

File: CHANGES.rst [hunk 5]
```diff
@@ -136,6 +136,7 @@ version number.  Breaking changes may occur at any point.
 
   - Processing [GH-793] [GH-812]
   - Racket [GH-799] [GH-873]
+  - Pug -- The pattern has been changed and this is not backwards compatible to Jade. [GH-951] [GH-1084]
```
Comment (lunaryorn) [inline CHANGES.rst]: Please move this to the latest version and remove the explanatory sentence.  It's enough for users to see that Jade was removed.

**Comment by lunaryorn** [comment]
@robbyoconnor  Thanks a lot for your contribution.

Please don't keep the jade checker, though.  If `jade` is deprecated there's no reason for us to keep it.  Users that still depend on jade should either stick to an older version of Flycheck or manually resurrect the jade checker in their configuration.

There are also a couple of minor issues in the PR.  Please address these and rebase onto master.

Thanks 👍

File: CHANGES.rst (hunk 4 above)
Comment (robbyoconnor) [inline CHANGES.rst]: Okay.

File: CHANGES.rst (hunk 5 above)
Comment (robbyoconnor) [inline CHANGES.rst]: okay :)

File: test/flycheck-test.el (hunk 3 above)
Comment (robbyoconnor) [inline test/flycheck-test.el]: I forgot :)

File: flycheck.el (hunk 2 above)
Comment (robbyoconnor) [inline flycheck.el]: Okay :) This was the formatter's work actually...

File: Cask (hunk 1 above)
Comment (robbyoconnor) [inline Cask]: It failed for me...

File: test/flycheck-test.el (hunk 3 above)
Comment (robbyoconnor) [inline test/flycheck-test.el]: Handled this.

File: CHANGES.rst (hunk 5 above)
Comment (robbyoconnor) [inline CHANGES.rst]: Done.

**Comment by robbyoconnor** [comment]
All green! :green_heart:

**Comment by robbyoconnor** [comment]
I think leaving it as it is...people shouldn't be using Jade anymore. If they need it, they can add it back...nevermind on the idea of making it work for both.

**Comment by robbyoconnor** [comment]
Hold off -- found a bug

**Comment by lunaryorn** [comment]
@robbyoconnor No offense meant, but jumping back and forth between "merge as is" and "hold off" is a bit fast for me :blush:  I'll unsubscribe from this PR to prevent "notifications overflow" :wink: 

Please ping me or @flycheck/maintainers when you're done to request another review.

**Comment by robbyoconnor** [comment]
Ready.

**Comment by robbyoconnor** [comment]
@flycheck/maintainers -- Ready for review!

**Comment by robbyoconnor** [comment]
# Syntax error
image

# error with extends/includes
image

# Runtime error 
image

File: CHANGES.rst [hunk 6]
```diff
@@ -8,11 +8,13 @@
     own ``.luacheckrc`` detection. Therefore ``flycheck-luacheckrc`` is
     no longer used [GH-1057]
   - ``:modes`` is now mandatory for syntax checker definitions [GH-1071]
+  - Remove jade checker [GH-951] [GH-1084]
 
 - New syntax checkers:
 
   - Elixir with ``dogma`` [GH-969]
   - sass and scss with ``sass-lint`` [GH-1070]
+  - pug [GH-951] [GH-1084]
```
Comment (cpitclaudel) [inline CHANGES.rst]: This should probably be either capitalized or wrapped in backticks, no?

File: flycheck.el [hunk 7]
```diff
@@ -7777,6 +7764,35 @@ See https://github.com/processing/processing/wiki/Command-Line"
...
+See URL `https://pugjs.org/'."
+  :command ("pug" "-p" (eval (expand-file-name (buffer-file-name))))
+  :standard-input t
+  :error-patterns
+  ;; errors with includes/extends (e.g., missing files)
+  ((error "Error: " (message) (zero-or-more not-newline) "\n"
+          (zero-or-more not-newline) "at "
+          (zero-or-more not-newline) " line " line)
+
+   ;; syntax/runtime errors
```
Comment (cpitclaudel) [inline flycheck.el]: This scares me a bit: what kind of errors do you have in mind when you mention "runtime errors"?

Comment (robbyoconnor) [inline flycheck.el]: Let me fix that wording.

File: CHANGES.rst (hunk 6 above)
Comment (robbyoconnor) [inline CHANGES.rst]: I can fix that.

File: flycheck.el (hunk 7 above)
Comment (robbyoconnor) [inline flycheck.el]: runtime errors (e.g., type errors -- which aren't syntax errors)

**Comment by robbyoconnor** [comment]
Sorry(triggering a CI build)

**Comment by robbyoconnor** [comment]
Build failures aren't my fault if they occur...seems travis times out

File: flycheck.el (hunk 7 above)
Comment (robbyoconnor) [inline flycheck.el]: # runtime error
```TypeError: /home/rob/flycheck/test/resources/language/pug/pug-runtime-error.pug:5
    3|   head
    4|     body
  > 5|     = foo.bar
    6| 

Cannot read property 'bar' of undefined
    at eval (eval at wrap (/home/rob/.nvm/versions/node/v5.12.0/lib/node_modules/pug-cli/node_modules/pug-runtime/wrap.js:6:10), <anonymous>:13:60)
    at template (eval at wrap (/home/rob/.nvm/versions/node/v5.12.0/lib/node_modules/pug-cli/node_modules/pug-runtime/wrap.js:6:10), <anonymous>:13:137)
    at ReadStream.<anonymous> (/home/rob/.nvm/versions/node/v5.12.0/lib/node_modules/pug-cli/index.js:230:20)
    at emitNone (events.js:85:20)
    at ReadStream.emit (events.js:179:7)
    at endReadableNT (_stream_readable.js:913:12)
    at _combinedTickCallback (internal/process/next_tick.js:74:11)
    at process._tickCallback (internal/process/next_tick.js:98:9)```

Comment (robbyoconnor) [inline flycheck.el]: Not really a syntax error, more of a type error :)

**Comment by robbyoconnor** [comment]
The Travis build failure is due to travis timing out

File: flycheck.el [hunk 8]
```diff
@@ -7777,6 +7764,35 @@ See https://github.com/processing/processing/wiki/Command-Line"
...
+
+See URL `https://pugjs.org/'."
+  :command ("pug" "-p" (eval (expand-file-name (buffer-file-name))))
+  :standard-input t
+  :error-patterns
+  ;; errors with includes/extends (e.g., missing files)
+  ((error "Error: " (message) (zero-or-more not-newline) "\n"
+          (zero-or-more not-newline) "at "
+          (zero-or-more not-newline) " line " line)
+
```
Comment (lunaryorn) [inline flycheck.el]: Please remove this blank line.

File: flycheck.el [hunk 9]
```diff
@@ -7777,6 +7764,35 @@ See https://github.com/processing/processing/wiki/Command-Line"
...
+                           (zero-or-more not-newline) "\n")
+                       (zero-or-more "-")  (zero-or-more not-newline) "|"
+                       (zero-or-more not-newline) "\n")
+          (zero-or-more not-newline) "\n"
+          (one-or-more
+           (zero-or-more not-newline) "|"
+           (zero-or-more not-newline) "\n") (zero-or-more not-newline) "\n"
+           (message) line-end))
+
+
```
Comment (lunaryorn) [inline flycheck.el]: Please remove this blank lines as well.

File: test/flycheck-test.el [hunk 10]
```diff
@@ -3547,6 +3543,17 @@ Why not:
    "language/processing/syntax_error/syntax_error.pde" 'processing-mode
    '(4 2 error "Syntax error, maybe a missing semicolon?"
        :checker processing)))
+(flycheck-ert-def-checker-test pug pug nil
```
Comment (lunaryorn) [inline test/flycheck-test.el]: Please add a blank line above.

File: test/flycheck-test.el [hunk 11]
```diff
@@ -3547,6 +3543,17 @@ Why not:
...
+  (flycheck-ert-should-syntax-check
+   "language/pug/pug.pug" 'pug-mode
+   '(2 1 error "unexpected token \"indent\"" :checker pug))
+  (flycheck-ert-should-syntax-check
```
Comment (lunaryorn) [inline test/flycheck-test.el]: Please split the test into multiple test cases.  There should only be one `-ert-should-syntax-check` call per `-ert-def-checker-test`.

The 3rd argument—which is `nil` currently—let's you change give a name to each checker test.

File: test/resources/language/pug.pug [hunk 12]
```diff
@@ -1,6 +1,6 @@
 doctype html
  html
   head
-    title Jade Examples
+    title Pug Examples
```
Comment (lunaryorn) [inline test/resources/language/pug.pug]: Please move this file into the `pug/` directory to put it beneath all the other pug files.

File: flycheck.el [hunk 13]
```diff
@@ -7372,19 +7372,6 @@ See URL `https://github.com/htacg/tidy-html5'."
...
-See URL `http://jade-lang.com'."
-  :command ("jade")
-  :standard-input t
-  :error-patterns
-  ((error line-start
-          "Error: Jade:" line (zero-or-more not-newline) "\n"
-          (one-or-more (and (zero-or-more not-newline) "|"
-                            (zero-or-more not-newline) "\n"))
-          (zero-or-more not-newline) "\n" (message) line-end))
-  :modes jade-mode)
```
Comment (lunaryorn) [inline flycheck.el]: Please remove this blank line.

File: test/resources/language/pug.pug (hunk 12 above)
Comment (robbyoconnor) [inline test/resources/language/pug.pug]: I thought i did... weird

File: flycheck.el (hunk 13 above)
Comment (robbyoconnor) [inline flycheck.el]: What? There is no blank line here...

**Comment by robbyoconnor** [comment]
Made the changes.

File: flycheck.el (hunk 13 above)
Comment (lunaryorn) [inline flycheck.el]: You removed the code only but no adjacent blank line, so there were to blank lines between the previous and the next checker now.  Looks fixed in the current diff.

File: flycheck.el [hunk 14]
```diff
@@ -7777,6 +7763,32 @@ See https://github.com/processing/processing/wiki/Command-Line"
...
+
+See URL `https://pugjs.org/'."
+  :command ("pug" "-p" (eval (expand-file-name (buffer-file-name))))
+  :standard-input t
+  :error-patterns
+  ;; errors with includes/extends (e.g., missing files)
+  ((error "Error: " (message) (zero-or-more not-newline) "\n"
+          (zero-or-more not-newline) "at "
+          (zero-or-more not-newline) " line " line)
+   ;; syntax/runtime errors(e.g, type errors)
```
Comment (lunaryorn) [inline flycheck.el]: There's a space missing before the opening parenthesis, and the comma after `e.g` should be a dot :)

File: flycheck.el [hunk 15]
```diff
@@ -7777,6 +7763,32 @@ See https://github.com/processing/processing/wiki/Command-Line"
...
   :predicate (lambda () (buffer-file-name)))
 
+(flycheck-define-checker pug
+  "A Pug syntax checker using the pug compiler.
+
+See URL `https://pugjs.org/'."
+  :command ("pug" "-p" (eval (expand-file-name (buffer-file-name))))
+  :standard-input t
+  :error-patterns
+  ;; errors with includes/extends (e.g., missing files)
```
Comment (lunaryorn) [inline flycheck.el]: There's a redundant comma

File: test/flycheck-test.el [hunk 16]
```diff
@@ -3548,6 +3544,22 @@ Why not:
...
+(flycheck-ert-def-checker-test pug pug syntax-error
+  (flycheck-ert-should-syntax-check
+   "language/pug/pug.pug" 'pug-mode
+   '(2 1 error "unexpected token \"indent\"" :checker pug)))
+
+(flycheck-ert-def-checker-test pug pug include-extends-error
+  (flycheck-ert-should-syntax-check
+   "language/pug/pug-extends.pug" 'pug-mode
+   '(1 nil error
+       "the \"basedir\" option is required to use includes and extends with \"absolute\" paths" :checker pug)))
```
Comment (lunaryorn) [inline test/flycheck-test.el]: Please move the message to the line before, and instead break before `:checker`.  That's how we did it in the other tests, I find that it's easier to read this way.

**Comment by lunaryorn** [comment]
LGTM, because LGTM.co didn't seem to catch the review comment.

**Comment by robbyoconnor** [comment]
Additional changes have been made.

**Comment by robbyoconnor** [comment]
CI failures are (again) not my fault -- ran tests local and they passed

**Comment by lunaryorn** [comment]
Many many thanks 👍

**Pull request merged**